main/lsp/requests/workspace_symbols.cc

```diff
@@ -123,61 +161,97 @@ vector<unique_ptr<SymbolInformation>> SymbolMatcher::doQuery(string_view query_v
-        auto [ownerSymbolIndex, ownerScore, ownerQueryIter] = scoreInfos[symbolData->owner._id];
-        optional<int> bestScore = nullopt;
-        if (scoreInfo.queryIter == queryEnd && !(ownerQueryIter == queryEnd && ownerScore <= scoreInfo.score)) {
-            bestScore = scoreInfo.score;
-        }
-        auto shortName = symbolData->name.data(gs)->shortName(gs);
-        auto [score, queryIter] = partialMatchSymbol(shortName, queryBegin, queryEnd, false);
-        if (queryIter == queryEnd && (!bestScore.has_value() || *bestScore > score)) {
-            bestScore = score;
-        }
-        if (ownerQueryIter != nullptr && ownerQueryIter != queryBegin && ownerQueryIter != queryEnd) {
-            auto [score, queryIter] = partialMatchSymbol(shortName, ownerQueryIter, queryEnd, false);
-            if (queryIter == queryEnd && (!bestScore.has_value() || *bestScore > score)) {
-                bestScore = ownerScore + score;
+    {
+        Timer timeit(gs.tracer(), "SymbolMatcher::doQuery::pass2");
+        for (u4 symbolIndex = 1; symbolIndex < gs.symbolsUsed(); symbolIndex++) {
+            auto symbolRef = core::SymbolRef(gs, symbolIndex);
+            auto symbolData = symbolRef.data(gs);
+            auto nameData = symbolData->name.data(gs);
+            if (!isEligibleSymbol(gs, symbolData, nameData)) {
+                continue;
+            }
+            auto shortName = nameData->shortName(gs);
+            optional<uint> bestScore = nullopt;
+            auto partialMatch = partialMatchSymbol(shortName, queryBegin, queryEnd, false, ceilingScore);
+            if (partialMatch.matchEnd == queryEnd) {
+                bestScore = partialMatch.score;
+            }
+            for (auto previousAncestorRef = symbolRef, ancestorRef = symbolData->owner;
+                 previousAncestorRef != ancestorRef && ancestorRef.exists();
+                 previousAncestorRef = ancestorRef, ancestorRef = ancestorRef.data(gs)->owner) {
+                auto [ancestorScore, ancestorEnd] = partialMatches[ancestorRef._id];
+                if (ancestorEnd == queryBegin || ancestorEnd == nullptr) {
+                    break; // no further ancestor will be of any help
+                }
+                if (ancestorEnd == queryEnd) {
+                    continue; // ancestor matched everything, so skip to its parent
+                }
+                auto bestOrCeilingScore = bestScore.value_or(ceilingScore);
+                if (ancestorScore >= bestOrCeilingScore) {
+                    continue; // matching this ancestor would be worse
+                }
+                auto [plusScore, plusEnd] =
+                    partialMatchSymbol(shortName, ancestorEnd, queryEnd, false, bestOrCeilingScore - ancestorScore);
+                if (plusEnd == queryEnd) {
+                    bestScore = min(bestOrCeilingScore, ancestorScore + plusScore);
+                }
+            }
+            if (bestScore.has_value()) {
```

Comment:
If `bestScore` isn't optional, you can check `if (bestScore < ceilingScore) {` instead (assuming `ceilingScore` is *exclusive*, i.e. you reject items with score == ceilingScore).